### PR TITLE
Pls review server and client implementation

### PR DESCRIPTION
**Pull request by bejelith**

### CODE REVIEW

File: ca/ssl.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:     x509.KeyUsageDigitalSignature,
+		DNSNames:     []string{"localhost", "example.com", "127.0.0.1"},
+	}
+
+	pk, _ := getRSA()
+	cert_b, _ := x509.CreateCertificate(rand.Reader, cert, ca, &pk.PublicKey, caPrivKey)
+
+	// Public key
+	certOut, _ := os.Create("server.crt")
```
Comment (rcanderson23) [inline ca/ssl.go:99]: Why are we ignoring errors here?

File: cmd/client/main.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	client := pb.NewSchedulerClient(conn)
+
+	signal.SetupSignalHandler(func(s os.Signal) {
+		conn.Close()
+	})
+	return client
+}
+
+func buildGRPCClient() (*grpc.ClientConn, error) {
+	ca_b, _ := os.ReadFile(*caFile)
```
Comment (rcanderson23) [inline cmd/client/main.go:180]: Probably should handle error here.

File: cmd/server/main.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	caPool := x509.NewCertPool()
+	if !caPool.AppendCertsFromPEM(data) {
+		return nil, fmt.Errorf("unable to append the CA certificate to CA pool")
+	}
+
+	tlsConfig := &tls.Config{
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		Certificates: []tls.Certificate{certificate},
+		ClientCAs:    caPool,
+	}
```
Comment (rcanderson23) [inline cmd/server/main.go:94]: RFD specifies usage of min version TLS 1.3 but is not specified here.

File: server/server.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	return &pb.GetResponse{Found: true, Pid: p.ID, Status: p.State}, nil
+}
+
+func (s *SchedulerServer) Start(ctx context.Context, r *pb.CreateRequest) (*pb.CreateResponse, error) {
+	var errorStr string
+	pid, err := s.Executor.Start(&executor.ProcessConfig{Cmd: r.Cmd, Args: r.Args})
+	if err != nil {
+		errorStr = err.Error()
+		return &pb.CreateResponse{Pid: pid, Error: &errorStr}, nil
+	}
```
Comment (rcanderson23) [inline server/server.go:30]: Feels a little weird to not return the error here.

File: ca/ssl.go (hunk 1 above)
Comment (bejelith) [reply ca/ssl.go:99]: its just a script so I didnt even much time on as on the challenge doc it was mentioned you have your own set of certificates. Ill make it more resilient

File: cmd/server/main.go (hunk 3 above)
Comment (bejelith) [reply cmd/server/main.go:94]: definitevely overloooked, thx for catching it

File: server/server.go (hunk 4 above)
Comment (bejelith) [reply server/server.go:30]: I was reluctant to return it as HTTP level error as the API call worked (TCP/HTTP level) and that is an error in the child itself so i think its more appropriate to return it as part of the response message.

For example, i find weird to return a "executable not found" as HTTP error. I could look into how docker does it or any suggestion is welcome :D

Comment (bejelith) [reply server/server.go:30]: If you look at https://grpc.io/docs/guides/error/ "richer error mode" pretty much describes my approach, the application error is passed as part of the GRPC protocol in the Response object

File: server/rbac.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+		if !i.AuthorizeCmd(role, r.GetCmd()) {
+			log.Warn("user unauthorized", "user", user, "role", role, "cmd", r.GetCmd())
+			return nil, fmt.Errorf("user %s/%s not authorized to run %s", role, user, r.GetCmd())
+		}
+		resp, e := handler(ctx, req)
+		pidResponse := resp.(PIDGetter)
+		i.AttributeOwnership(user, uint64(pidResponse.GetPid()))
+
+		//if log.Level() == log.LevelDebug {
+		//
```
Comment (rcanderson23) [inline server/rbac.go:104]: Unnecessary comments here and just below

File: server/rbac.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+		}
+	}
+	return i.userIsAdmin(user)
+}
+
+func (i *RBACInterceptor) userIsAdmin(user string) bool {
+	// Ultimately check if the user is admin
+	if role, exists := i.users[user]; !exists {
+		return false
+	} else {
```
Comment (rcanderson23) [inline server/rbac.go:170]: nit: unnecessary else

**Review by rcanderson23 — APPROVED** [review]

File: server/rbac.go (hunk 5 above)
Comment (bejelith) [reply server/rbac.go:104]: I believe i need to create a local instance of slog to do that, i put a comment just to say that on debug logging  i'd  put on if statement before it to skip processing the log string if not necessary

File: server/server.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+func (s *SchedulerServer) Stdout(r *pb.OutputRequest, stream pb.Scheduler_StdoutServer) error {
+	job := s.Executor.Get(r.Pid)
+	if job == nil {
+		return fmt.Errorf("job does not exists")
+	}
+	reader, _ := s.Executor.Stdout(job.ID)
+
+	var buffer = make([]byte, 1024)
+	for {
+		n, err := reader.Read(buffer)
```
Comment (sclevine) [inline server/server.go:45]: bytes should also be processed from `Read`, even if there's an error

Comment (bejelith) [reply server/server.go:46]: fixed

**Pull request merged**